nemo/collections/asr/modules/flashlight_decoder.py

```diff
@@ -0,0 +1,240 @@
+                beam_size_token=int(beam_size_token),
+                beam_threshold=beam_threshold,
+                lm_weight=lm_weight,
+                word_score=word_score,
+                unk_score=unk_weight,
+                sil_score=sil_weight,
+                log_add=False,
+                criterion_type=self.criterion_type,
+            )
+
+            self.decoder = LexiconDecoder(
+                self.decoder_opts, self.trie, self.lm, self.silence, self.blank, self.unk_word, [], self.unit_lm,
+            )
+        else:
+            assert self.unit_lm, "lexicon free decoding can only be done with a unit language model"
+            from flashlight.lib.text.decoder import LexiconFreeDecoder, LexiconFreeDecoderOptions
+
+            d = {w: [[w]] for w in self.tokenizer.vocab + ([] if '<unk>' in self.tokenizer.vocab else ['<unk>'])}
+            self.word_dict = create_word_dict(d)
+            self.lm = KenLM(lm_path, self.word_dict)
+            self.decoder_opts = LexiconFreeDecoderOptions(
+                beam_size=beam_size,
+                beam_size_token=int(beam_size_token),
+                beam_threshold=beam_threshold,
+                lm_weight=lm_weight,
+                sil_score=sil_weight,
+                log_add=False,
+                criterion_type=self.criterion_type,
+            )
+            self.decoder = LexiconFreeDecoder(self.decoder_opts, self.lm, self.silence, self.blank, [])
+
+    def _get_tokens(self, idxs: List[int]):
+        """Normalize tokens by handling CTC blank, ASG replabels, etc."""
+
+        idxs = (g[0] for g in it.groupby(idxs))
+        idxs = filter(lambda x: x != self.blank and x != self.silence, idxs)
+
+        return torch.LongTensor(list(idxs))
+
+    def _get_timesteps(self, token_idxs: List[int]):
+        """Returns frame numbers corresponding to every non-blank token.
+        Parameters
+        ----------
+        token_idxs : List[int]
+            IDs of decoded tokens.
+        Returns
+        -------
+        List[int]
+            Frame numbers corresponding to every non-blank token.
+        """
+
+        timesteps = []
+        for i, token_idx in enumerate(token_idxs):
+            if token_idx == self.blank:
+                continue
+            if i == 0 or token_idx != token_idxs[i - 1]:
+                timesteps.append(i)
+
+        return timesteps
+
+    # @typecheck(ignore_collections=True)
+    @torch.no_grad()
+    def forward(self, log_probs: Union[np.ndarray, torch.Tensor]):
+        if isinstance(log_probs, np.ndarray):
+            log_probs = torch.from_numpy(log_probs).float()
+        if log_probs.dim() == 2:
+            log_probs = log_probs.unsqueeze(0)
+
+        emissions = log_probs.cpu().contiguous()
+
+        B, T, N = emissions.size()
+        hypos = []
+        for b in range(B):
+            emissions_ptr = emissions.data_ptr() + 4 * b * emissions.stride(0)
+            results = self.decoder.decode(emissions_ptr, T, N)

```

Comment:
Yes, that is how the upstream model returns the information, with some sort of padding (I presume it's the blank or silence token, you can see in the model code itself). Since we specify both of those tokens to the flashlight module (see above in the code), they are naturally removed anyway, so it doesn't bother the decoding at all that they're all the same length and padded.